Sources/AWSXRayInstrument/Instrument.swift

```diff
@@ -0,0 +1,30 @@
+import AWSXRayRecorder
+import Baggage
+import Dispatch
+import Instrumentation
+
+extension XRayRecorder: TracingInstrument {
+    public func extract<Carrier, Extractor>(_ carrier: Carrier, into baggage: inout BaggageContext, using extractor: Extractor) where Carrier == Extractor.Carrier, Extractor: ExtractorProtocol {
+        // TODO: impl
+    }
+
+    public func inject<Carrier, Injector>(_ baggage: BaggageContext, into carrier: inout Carrier, using injector: Injector) where Carrier == Injector.Carrier, Injector: InjectorProtocol {
+        // TODO: impl
+    }
+
+    public func startSpan(
+        named operationName: String,
+        context: BaggageContext,
+        ofKind kind: SpanKind,
+        at timestamp: DispatchTime?
+    ) -> Span {
+        // TODO: parse context, extend existing recorder/segment to use `Baggage`?
+        // TODO: map DispatchTime to Timestamp
+        // TODO: does kind has any meaning?
+        // TODO: what id the context is invalid (BaggageContext does not have a valid one),
+        // ^ typically as far XRay is concerned `AWS_XRAY_CONTEXT_MISSING` is used to configure how to handle it
+        // TracingInstrument should indicate how to handle that
```

Comment:
@slashmo ?

Reply:
I'm not sure tracing instrument needs to  indicate how to deal with that, various tracers can do various things.

> AWS_XRAY_CONTEXT_MISSING – Set to LOG_ERROR to avoid throwing exceptions when your instrumented code attempts to record data when no segment is open.
> 
> - RUNTIME_ERROR – Throw a runtime exception (default).
> - LOG_ERROR – Log an error and continue.

the throwing is an interesting question... It's a bit annoying to make those funcs throwing to be honest... WDYT?

Reply:
@ktoso

I do agree its extreme, its seems AWS does it consistently though, see https://github.com/awslabs/aws-xray-sdk-with-opentelemetry/blob/master/sdk/src/main/java/com/amazonaws/xray/opentelemetry/tracing/propagation/HttpTraceContext.java

```java
  @Override
  //TODO - Accept w3c headers
  public <C> SpanContext extract(final C carrier, final Getter<C> getter) {
    String xrayHeader = getter.get(carrier, XRAY_HEADER_NAME);
    if (xrayHeader != null) {
      return ContextUtils.headerToSpanContext(TraceHeader.fromString(xrayHeader));
    } else {
      throw new RuntimeException("Could not extract X-Ray Trace Header.");
    }
  }
```

Reply:
to be followed up in https://github.com/pokryfka/aws-xray-sdk-swift/issues/28#issue-665280618